File: csharp/src/detectors/xml-external-entity/XmlExternalEntity.cs
using System.Xml;

namespace XXEInjection
{
    class XmlExternalEntity
    {
        // {fact rule=xml-external-entity@v1.0 defects=1}
        public void XmlExternalEntityNoncompliant(string str)
        {
            XmlReaderSettings readersetting = new XmlReaderSettings();
            // Noncompliant: `DtdProcessing.Parse` enables DTD processing.
            readersetting.DtdProcessing = DtdProcessing.Parse;
            XmlReader reader = XmlReader.Create(new StringReader(str),readersetting);
            
            while (reader.Read())
            {
                Console.WriteLine(reader.Value);
            }
            Console.ReadLine();
        }
        // {/fact}
        
        // {fact rule=xml-external-entity@v1.0 defects=0}
        public void XmlExternalEntityCompliant(string str)
        {
            XmlReaderSettings readersetting = new XmlReaderSettings();
            // Compliant: `DtdProcessing.Ignore` disables DTD processing without warnings or exceptions.
            readersetting.DtdProcessing = DtdProcessing.Ignore;
            XmlReader reader = XmlReader.Create(new StringReader(str),readersetting);
            
            while (reader.Read())
            {
                Console.WriteLine(reader.Value);
            }
            Console.ReadLine();
        }
        // {/fact}
    }
}

File: typescript/src/detector/high/xml-external-entity/xml-external-entity.ts
// {fact rule=xml-external-entity@v1.0 defects=1}
import libxmljs from "libxmljs"
import fs from 'fs'

function xmlExternalEntityNoncompliant() {
    const xml = fs.readFileSync("foo.xml")
    // Noncompliant: sets `noent` to true which enables the parsing of external entities.
    const xmlDoc = libxmljs.parseXml(xml, { noent: true, noblanks: true })
}
// {/fact}


// {fact rule=xml-external-entity@v1.0 defects=0}
import libxmljs from "libxmljs"
import fs from 'fs'

function xmlExternalEntityCompliant() {
    const xml = fs.readFileSync("foo.xml")
    // Compliant: parsing of external entities is disabled by default.
    const xmlDoc = libxmljs.parseXml(xml, { noblanks: true })
}
// {/fact}
